lib/prefetch.js

```diff
@@ -0,0 +1,113 @@
+import React from 'react'
+import Link, { isLocal } from './link'
+import { parse as urlParse } from 'url'
+
+class Messenger {
+  constructor () {
+    this.id = 0
+    this.callacks = {}
+    this.serviceWorkerReadyCallbacks = []
+    navigator.serviceWorker.addEventListener('message', ({ data }) => {
+      if (data.action !== 'REPLY') return
+      if (this.callacks[data.replyFor]) {
+        this.callacks[data.replyFor](data)
+      }
+    })
+    this._register()
+  }
+
+  send (payload, cb) {
+    if (this.serviceWorkerState === 'REGISTERED') {
+      this._send(payload, cb)
+    } else {
+      this.serviceWorkerReadyCallbacks.push(() => {
+        this._send(payload, cb)
+      })
+    }
+  }
+
+  _send (payload, cb = () => {}) {
+    const id = this.id ++
+    const newPayload = { ...payload, id }
+
+    this.callacks[id] = (data) => {
+      if (data.error) {
+        cb(data.error)
+      } else {
+        cb(null, data.result)
+      }
+
+      delete this.callacks[id]
+    }
+
+    navigator.serviceWorker.controller.postMessage(newPayload)
+  }
+
+  _register () {
+    this.serviceWorkerState = 'REGISTERING'
+    navigator.serviceWorker.register('/_next-prefetcher.js')
```

Comment:
I tried that way. See this commit message. 
https://github.com/zeit/next.js/pull/375/commits/09325d9443ac09e5155fc509b53ceb35d66aa208